components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc

```diff
@@ -183,4 +253,104 @@ EngineConsumerConversationAPI::GetAssociatedContentConversationEvent(
+            // Return deduped topics from the response.
+            std::vector<EngineConsumer::GenerationResult> results = {result};
+            std::move(callback).Run(
+                EngineConsumerConversationAPI::
+                    GetStrArrFromTabOrganizationResponses(results));
+          },
+          std::move(callback)));
+}
+
+void EngineConsumerConversationAPI::ProcessTabChunks(
+    const std::vector<Tab>& tabs,
+    ConversationEventType event_type,
+    base::OnceCallback<void(std::vector<GenerationResult>)> merge_callback,
+    const std::string& topic) {
+  CHECK(event_type == ConversationEventType::GetSuggestedTopicsForFocusTabs ||
+        event_type == ConversationEventType::GetFocusTabsForTopic);
+
+  // Split tab into chunks of 75
+  size_t num_chunks = (tabs.size() + kChunkSize - 1) / kChunkSize;
+  const auto barrier_callback = base::BarrierCallback<GenerationResult>(
+      num_chunks, std::move(merge_callback));
+
+  for (size_t chunk = 0; chunk < num_chunks; ++chunk) {
+    base::Value::List tab_value_list;
+    for (size_t i = chunk * kChunkSize;
+         i < std::min((chunk + 1) * kChunkSize, tabs.size()); ++i) {
+      tab_value_list.Append(base::Value::Dict()
+                                .Set("id", tabs[i].id)
+                                .Set("title", tabs[i].title)
+                                .Set("url", tabs[i].origin.Serialize()));
+    }
+
+    std::vector<ConversationEvent> conversation;
+    conversation.push_back(
+        {mojom::CharacterType::HUMAN, event_type,
+         base::WriteJson(tab_value_list).value_or(std::string()), topic});
+
+    api_->PerformRequest(std::move(conversation), "" /* selected_language */,
+                         base::NullCallback() /* data_received_callback */,
+                         barrier_callback /* data_completed_callback */);
+  }
+}
+
+void EngineConsumerConversationAPI::MergeSuggestTopicsResults(
+    GetSuggestedTopicsCallback callback,
+    std::vector<GenerationResult> results) {
+  // Merge the result and send another request to dedupe topics.
+  DedupeTopics(GetStrArrFromTabOrganizationResponses(results),
+               std::move(callback));
+}
+
+void EngineConsumerConversationAPI::GetSuggestedTopics(
+    const std::vector<Tab>& tabs,
+    GetSuggestedTopicsCallback callback) {
+  ProcessTabChunks(
+      tabs, ConversationEventType::GetSuggestedTopicsForFocusTabs,
+      base::BindOnce(&EngineConsumerConversationAPI::MergeSuggestTopicsResults,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)),
+      "" /* topic */);
+}
+
+void EngineConsumerConversationAPI::GetFocusTabs(
+    const std::vector<Tab>& tabs,
+    const std::string& topic,
+    EngineConsumer::GetFocusTabsCallback callback) {
+  ProcessTabChunks(tabs, ConversationEventType::GetFocusTabsForTopic,
+                   base::BindOnce(
+                       [&](EngineConsumer::GetFocusTabsCallback callback,
```

Comment:
Should we be passing in a WeakPtr here, instead of binding the ambient context?

Reply:
https://github.com/brave/brave-core/pull/28103/commits/3cc2d32e22a3b235dbbd54a493302eb68ca40251